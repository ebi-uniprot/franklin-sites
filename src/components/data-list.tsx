/* eslint-disable react/no-unused-prop-types */
import { memo, ReactNode, HTMLAttributes } from 'react';
import cn from 'classnames';

import Loader from './loader';

import useDataCheckboxes from '../hooks/useDataCheckboxes';

import withDataLoader, { WrapperProps } from './data-loader';

export type Props<Datum> = {
  /**
   * The data to be displayed
   */
  data: Datum[];
  /**
   * Flag saying that data is loading, so we might be showing stale data
   */
  loading?: boolean;
  /**
   * A function that returns a unique ID for each of the data objects.
   * Same function signature as a map function.
   */
  getIdKey: (datum: Datum, index: number, data: Datum[]) => string;
  /**
   * A callback that is called whenever a user selects or unselects a row.
   */
  onSelectionChange?: (event: MouseEvent | KeyboardEvent) => void;
  /**
   * A renderer function for each item of the list.
   * Make sure that it doesn't change unecessarily by wrapping it in useCallback
   */
  dataRenderer: (datum: Datum) => ReactNode;
};

type BasicDatum = Record<string, unknown>;

type DataListItemProps<Datum> = {
  datum: Datum;
  id: string;
  dataRenderer: (datum: Datum) => ReactNode;
  loading: boolean;
  firstItem: boolean;
};

const DataListItem = <Datum extends BasicDatum>({
  datum,
  id,
  dataRenderer,
  loading,
  firstItem,
}: DataListItemProps<Datum>) => {
  let rendered: ReactNode;
  try {
    rendered = dataRenderer(datum);
  } catch (error) {
    /**
     * We get here only if the renderer fails. If the renderer returns null of
     * undefined because of a lack of data, then it will no throw and will not
     * display the loader at all
     */
    if (!loading) {
      throw error;
    } else {
      rendered = firstItem && <Loader />;
    }
  }

  return <li key={id}>{rendered}</li>;
};
const MemoizedDataListItem = memo(DataListItem) as typeof DataListItem;

export const DataList = <Datum extends BasicDatum>({
  data,
  getIdKey,
  dataRenderer,
  loading = false,
  onSelectionChange,
  className,
  ...props
}: Props<Datum> & HTMLAttributes<HTMLUListElement>) => {
  const { checkboxContainerRef } = useDataCheckboxes(onSelectionChange);

  return (
    <ul
      {...props}
      className={cn('data-list', 'no-bullet', className)}
      ref={checkboxContainerRef}
    >
      {data.map((datum, index) => {
        const id = getIdKey(datum, index, data);
        return (
          <MemoizedDataListItem
            key={id}
            datum={datum}
            id={id}
            dataRenderer={dataRenderer}
            loading={loading}
            firstItem={index === 0}
          />
        );
      })}
    </ul>
  );
};

export const DataListWithLoader = <Datum extends BasicDatum>(
  props: WrapperProps<Datum> & Props<Datum> & HTMLAttributes<HTMLElement>
) => <>{withDataLoader<Datum, typeof props>(DataList)(props)}</>;
