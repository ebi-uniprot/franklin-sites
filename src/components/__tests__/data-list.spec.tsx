import { render, fireEvent, screen } from '@testing-library/react';

import { DataListWithLoader as DataList } from '../data-list';

describe('DataList', () => {
  const data = Array.from({ length: 10 }, (_, index) => ({
    id: `id${index}`,
    content: index,
  }));

  let onLoadMoreItems: () => void;

  const renderList = ({ hasMoreData = true, clickToLoad = true } = {}) =>
    render(
      <DataList
        onLoadMoreItems={onLoadMoreItems}
        hasMoreData={hasMoreData}
        data={data}
        getIdKey={(datum) => datum.id}
        clickToLoad={clickToLoad}
        dataRenderer={(item) => <p>{item.content}</p>}
      />
    );

  beforeEach(() => {
    onLoadMoreItems = jest.fn();
  });

  it('should render autoload', () => {
    const { asFragment } = renderList({ clickToLoad: false });
    expect(asFragment()).toMatchSnapshot();
  });

  it('should render click-to-load', () => {
    const { asFragment } = renderList();
    expect(asFragment()).toMatchSnapshot();
  });

  it('should request more data', () => {
    renderList();
    expect(onLoadMoreItems).not.toHaveBeenCalled();
    const clickToLoadMore = screen.getByTestId('click-to-load-more');
    fireEvent.click(clickToLoadMore);
    expect(onLoadMoreItems).toHaveBeenCalled();
  });

  it('should not show the option to load more data', () => {
    renderList({ hasMoreData: false });
    const clickToLoadMore = screen.queryByTestId('click-to-load-more');
    expect(clickToLoadMore).toBeNull();
  });
});
