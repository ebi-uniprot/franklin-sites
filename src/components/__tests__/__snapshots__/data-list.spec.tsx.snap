// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DataList should render autoload 1`] = `
<DocumentFragment>
  <div
    class="data-list"
  >
    <section>
      <p>
        0
      </p>
    </section>
    <section>
      <p>
        1
      </p>
    </section>
    <section>
      <p>
        2
      </p>
    </section>
    <section>
      <p>
        3
      </p>
    </section>
    <section>
      <p>
        4
      </p>
    </section>
    <section>
      <p>
        5
      </p>
    </section>
    <section>
      <p>
        6
      </p>
    </section>
    <section>
      <p>
        7
      </p>
    </section>
    <section>
      <p>
        8
      </p>
    </section>
    <section>
      <p>
        9
      </p>
    </section>
  </div>
  <div
    class="data-loader__loading"
  >
    <button
      class="button secondary"
      data-testid="click-to-load-more"
      type="button"
    >
      Load more data
    </button>
  </div>
</DocumentFragment>
`;

exports[`DataList should render click-to-load 1`] = `
<DocumentFragment>
  <div
    class="data-list"
  >
    <section>
      <p>
        0
      </p>
    </section>
    <section>
      <p>
        1
      </p>
    </section>
    <section>
      <p>
        2
      </p>
    </section>
    <section>
      <p>
        3
      </p>
    </section>
    <section>
      <p>
        4
      </p>
    </section>
    <section>
      <p>
        5
      </p>
    </section>
    <section>
      <p>
        6
      </p>
    </section>
    <section>
      <p>
        7
      </p>
    </section>
    <section>
      <p>
        8
      </p>
    </section>
    <section>
      <p>
        9
      </p>
    </section>
  </div>
  <div
    class="data-loader__loading"
  >
    <button
      class="button secondary"
      data-testid="click-to-load-more"
      type="button"
    >
      Load more data
    </button>
  </div>
</DocumentFragment>
`;
